intellij-plugin-verifier/verifier-test/src/test/java/com/jetbrains/pluginverifier/tests/BaseBytecodeTest.kt

```diff
@@ -0,0 +1,394 @@
+    val className = dynamicType.typeDescription.simpleName
+    file("$className.class", dynamicType.bytes)
+  }
+
+  @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
+  private fun load(classDynamicType: DynamicType.Unloaded<Object>, classLoader: ClassLoader): Class<out Any> {
+    return classDynamicType.load(classLoader, ClassLoadingStrategy.Default.WRAPPER).loaded
+  }
+
+  protected fun String.construct() = byteBuddy
+    .subclass(Object::class.java)
+    .name(this)
+
+  @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
+  protected fun String.constructWithMethod(
+    method: String,
+    returnType: Type,
+    modifier: Int = Modifier.PUBLIC
+  ): DynamicType.Builder.MethodDefinition.ParameterDefinition.Initial<Object> {
+    return construct()
+      .defineMethod(method, returnType, modifier)
+  }
+
+  protected fun buildIdeWithBundledPlugins(
+    includeKotlinStdLib: Boolean = false,
+    javaPluginClassesBuilder: (ContentBuilder).() -> Unit
+  ): Ide {
+    val ideaDirectory = buildDirectory(temporaryFolder.newFolder("idea").toPath()) {
+      file("build.txt", "IU-192.1")
+      dir("lib") {
+        zip("idea.jar") {
+          dir("META-INF") {
+            file("plugin.xml") {
+              """
+                <idea-plugin>
+                  <id>com.intellij</id>
+                  <name>IDEA CORE</name>
+                  <version>1.0</version>
+                  <module value="com.intellij.modules.all"/>                
+                </idea-plugin>
+                """.trimIndent()
+            }
+          }
+        }
+        if (includeKotlinStdLib) {
+          findKotlinStdLib().apply {
+            file(simpleName, this)
+          }
+        }
+      }
+      dir("plugins") {
+        dir("java") {
+          dir("lib") {
+            zip("java.jar") {
+              dir("META-INF") {
+                file("plugin.xml") {
+                  """
+                    <idea-plugin>
+                      <id>com.intellij.java</id>
+                      <module value="com.intellij.modules.java"/>
+                    </idea-plugin>
+                    """.trimIndent()
+                }
+              }
+
+              //Generate content of Java plugin.
+              javaPluginClassesBuilder()
+            }
+          }
+        }
+      }
+    }
+
+    // Fast assert IDE is fine
+    val ide = IdeManager.createManager().createIde(ideaDirectory)
+    assertEquals("IU-192.1", ide.version.asString())
+
+    val javaPlugin = ide.bundledPlugins.find { it.pluginId == "com.intellij.java" }!!
+    assertEquals("com.intellij.java", javaPlugin.pluginId)
+    assertEquals(setOf("com.intellij.modules.java"), javaPlugin.definedModules)
+
+    return ide
+  }
+
+  /**
+   * Builds an instance of the IDE with specified bundled plugins.
+   *
+   * @param bundledPlugins List of plugins to include in the `plugins` directory.
+   * @param bundledCorePlugins List of plugins to include in the `lib` directory of the IDE.
```

Comment:
what are "plugins in lib directory of the IDE" ??

Reply:
Improved wording.

Note that the current testing infrastructure does not quite fully support mock plugins in mock IDE's `plugins` directory. This is a workaround for such scenario.